pyuploadtool/releases_hosting_provider/github_releases.py

```diff
@@ -77,6 +78,13 @@ def create_release(self, metadata: ReleaseMetadata, artifacts):
 
         message = f"Build log: {metadata.build_log_url}"
 
+        if os.getenv("GENERATE_CHANGELOG", "").lower() == "true":
+            github_chglog = GitHubChangelogGenerator(
```

Comment:
Not a fan of how this is integrated, as it mixes the "metadata gathering" with a use of metadata.

Changelogs are actually a specific type of metadata. You should probably extend the metadata class with a new property. Then, the GitHub actions builder can generate the metadata and fill it into the metadata object. Finally, the GitHub releases hosting provider can read those data (if available) and update the description.

Reply:
I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional. 
I do not understand the concept of optional metadata, in this case
* we are not forcing all the users to accept the changelog we generated
* this changelog feature, is currently bound to a single infrastructure.. GitHub Releases

Until we add the support for other platforms, I prefer to keep as is.

Reply:
> I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional.
I do not understand the concept of optional metadata, in this case

How do _you_ define metadata? Metadata here is defined as "some kind of data that can be used to enhance the uploading of releases". Of course changelogs are a sort of metadata.

> I prefer to keep metadata static, and not optional.

It _is_ static (as the commit data do not change once they're calculated). I don't know how to interpret optional. To me, all metadata _can_ be used by release hosting providers. Some is used, some isn't, some is mandatory.

> this changelog feature, is currently bound to a single infrastructure.. GitHub Releases

Adding it to the metadata doesn't mean a release hosting provider _must_ use it. As explained, metadata _may_ be processed. The idea is to gather as much as possible, then present it to the hosting providers, which _can_ but do not have to make any use of it. It's up to the release hosting provider to use any of the data. That adds a _lot_ of flexibility to the code.

> Until we add the support for other platforms, I prefer to keep as is.

This version goes against the underlying architecture of the software, unfortunately.

Reply:
> > I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional.
> > I do not understand the concept of optional metadata, in this case
> 
> How do _you_ define metadata? Metadata here is defined as "some kind of data that can be used to enhance the uploading of releases". Of course changelogs are a sort of metadata.
> 
> > I prefer to keep metadata static, and not optional.
> 
> It _is_ static (as the commit data do not change once they're calculated). I don't know how to interpret optional. To me, all metadata _can_ be used by release hosting providers. Some is used, some isn't, some is mandatory.
> 

The  question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit

> > this changelog feature, is currently bound to a single infrastructure.. GitHub Releases
> 
> Adding it to the metadata doesn't mean a release hosting provider _must_ use it. As explained, metadata _may_ be processed. The idea is to gather as much as possible, then present it to the hosting providers, which _can_ but do not have to make any use of it. It's up to the release hosting provider to use any of the data. That adds a _lot_ of flexibility to the code.
> 
> > Until we add the support for other platforms, I prefer to keep as is.
> 
> This version goes against the architecture of the software, unfortunately.

Ok.. temporily, I will add `metadata.changelog` which would be `None` until the user requests changelog. I hope that's fine.

Reply:
> Until we add the support for other platforms, I prefer to keep as is.

This is the wrong approach. This would be duplicated in any other release hosting provider. In fact, I intend to extend the WebDAV provider with this functionality as well. Do you think it is a good idea to duplicate the code?

The reason I implemented the software architecture as it is is to avoid such scenarios. The code simply doesn't belong here.

> The question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit

You won't run into rate limits easily on GitHub actions. Check how the `github` package is used everywhere. It's using the `$GITHUB_TOKEN` environment variable. These tokens are very unlike to run into rate limits.

> Ok.. temporily, I will add metadata.changelog which would be None until the user requests changelog. I hope that's fine.

That's how the metadata object works. I should probably document that system a bit better. Why would this be temporarily, though?

Reply:
> > Until we add the support for other platforms, I prefer to keep as is.
> 
> This is the wrong approach. This would be duplicated in any other release hosting provider. In fact, I intend to extend the WebDAV provider with this functionality as well. Do you think it is a good idea to duplicate the code?
> 
> The reason I implemented the software architecture as it is is to avoid such scenarios. The code simply doesn't belong here.

Ah I see
> 
> > The question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit
> 
> You won't run into rate limits easily on GitHub actions. Check how the `github` package is used everywhere. It's using the `$GITHUB_TOKEN` environment variable. These tokens are very unlike to run into rate limits.
> 
> > Ok.. temporily, I will add metadata.changelog which would be None until the user requests changelog. I hope that's fine.
> 
> That's how the metadata object works. I should probably document that system a bit better. Why would this be temporarily, though?

Oh, I thought you wanted me to make another class, which should be of `ReleaseMetadata` type. Now it makes sense.